### PR TITLE
Complete libsnark sha integration

### PR DESCRIPTION
**Pull request by Schaeff**
Due to a wrong git workflow (decoupled a PR that applied unrelated changes the wrong way, by branching off it and reverting some changes with commits rather than cherry-picking), some of the changes to introduce the libsnark sha256 integration were not applied.

This PR adds all changes which were lacking in #66 and resolves conflicts with develop.

### CODE REVIEW

**Comment by Schaeff** [comment]
Identified an issue with the setup. On it.

**Comment by Kyroy** [comment]
If there is an issue, you should add a test ;)
Because tests were successful :D

**Comment by Schaeff** [comment]
It's true that the setup part needs more tests, it had been a black box for me so far.

File: src/executable.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+impl<T: Field> Executable<T> for Sha256Libsnark {
+	fn get_signature(&self) -> (usize, usize) {
+		(self.input_count, self.output_count)
+	}
+	fn execute(&self, inputs: &Vec<T>) -> Result<Vec<T>, String> {
+		let (expected_input_count, expected_output_count) = (self as &Executable<T>).get_signature();
+		assert!(inputs.len() == expected_input_count);
+		let witness: standard::Witness = serde_json::from_str(&get_sha256_witness(inputs)).unwrap();
```
Comment (JacobEberhardt) [inline src/executable.rs]: unwrap can be removed

File: src/libsnark.rs [hunk 2]
```diff
@@ -48,21 +54,80 @@ pub fn setup<T: Field> (
...
+      }
+      for &(idx, ref val) in &b[row] {
+          b_vec.push((row as i32, idx as i32, vec_as_u8_32_array(&val.into_byte_vector())));
+      }
+      for &(idx, ref val) in &c[row] {
+          c_vec.push((row as i32, idx as i32, vec_as_u8_32_array(&val.into_byte_vector())));
+      }
+    }
+
+    // Sizes and offsets in bytes for our struct {row, id, value}
```
Comment (JacobEberhardt) [inline src/libsnark.rs]: This could be fragile. Why is it not uints?

File: src/libsnark.rs [hunk 3]
```diff
@@ -48,21 +54,80 @@ pub fn setup<T: Field> (
...
+    let mut a_arr: Vec<u8> = vec![0u8; STRUCT_SIZE * a_vec.len()];
+    let mut b_arr: Vec<u8> = vec![0u8; STRUCT_SIZE * b_vec.len()];
+    let mut c_arr: Vec<u8> = vec![0u8; STRUCT_SIZE * c_vec.len()];
+    use std::mem::transmute;
+    for (id, (row, idx, val)) in a_vec.iter().enumerate() {
+      let row_bytes: [u8; ROW_SIZE] = unsafe { transmute(row.to_le()) };
+      let idx_bytes: [u8; IDX_SIZE] = unsafe { transmute(idx.to_le()) };
+
+      for x in 0..ROW_SIZE {
+        a_arr[id * 40 + x] = row_bytes[x];
```
Comment (JacobEberhardt) [inline src/libsnark.rs]: Replace by constant

File: src/standard.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+use executable::Sha256Libsnark;
+use parameter::Parameter;
+use reduce::Reduce;
+
+// for r1cs import, can be moved.
+// r1cs data strucutre reflecting JSON standard format:
+//{variables:["a","b", ... ],
+//constraints:[
+// [{offset_1:value_a1,offset2:value_a2,...},{offset1:value_b1,offset2:value_b2,...},{offset1:value_c1,offset2:value_c2,...}]
+//]}
```
Comment (JacobEberhardt) [inline src/standard.rs]: Outdated comment, standard is impl.

**Pull request merged**